### PR TITLE
Remove support for EndpointsLeases and ConfigMapsLeases lock from leader election

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #80289
Continuation of: #106852

```release-note-action-required
client-go: the leader election library no longer supports leader-election locks for `endpointsleases` or `configmapsleases` resources. Users of those lock resources should move to `leases` which is safe to do in a single release. Users of previously removed `endpoints` or `configmaps` lock should first transition to `endpointsleases` or `configmapsleases`, which are supported in k8s.io/client-go v0.17 - v0.27, before upgrading to version v0.28+ and switching to `leases`
```

/kind feature
/sig api-machinery
/priority important-soon

/assign @liggitt @deads2k

### CODE REVIEW

**Comment by liggitt** [comment]
removing this means that anyone still on configmap / endpoints election would have to jump to a 1.27.x level of client-go to use the transitional resource, then jump to 1.28+

is it important enough to drop support for the transitional resource to make consumers do both of those hops?

**Comment by wojtek-t** [comment]
> removing this means that anyone still on configmap / endpoints election would have to jump to a 1.27.x level of client-go to use the transitional resource, then jump to 1.28+

> is it important enough to drop support for the transitional resource to make consumers do both of those hops?


This is a good question and there isn't a clear answer to it:
- on one hand what you pointed out is a risk here and potentially additional work to some people
- on the other hand, we really want people to migrate to leases, (staying on multilock which is now possible also isn't great, as they are obviously less reliable (we need to ensure two locks at the same time) and worse for performance)


When discussing it with @deads2k back then, we wanted to remove this support after 3 releases. In 1.28 there will be 4. But maybe we want to leave it for a bit more? I just don't think we should support it forever...

**Comment by deads2k** [comment]
> is it important enough to drop support for the transitional resource to make consumers do both of those hops?

Allowing consumers to continue using endpoints isn't zero cost.  Because we cannot determine the intent of using a given API in p&f, there is additional undesired contention that cannot be reliably resolved.  Without removing the capability in the library, I'm unsure how to eliminate the contention.

**Comment by wojtek-t** [comment]
+1 to David
"as they are obviously less reliable (we need to ensure two locks at the same time)" from my comment was supposed to catch also aspects of P&F too.

**Comment by liggitt** [comment]
that's fair... I guess the joint resource lock functionality has _long_ been supported in client-go, so this would mean a client-go consumer needing to transition would need to roll out a release on a client-go version 0.17.0 - 0.27.0 and use endpointsleases / configmapsleases before updating to 0.28.0+

Update the release note to describe what people using endpoints or configmaps as their lock resource should do (the specific versions of client-go that support the joint lock functionality).

**Comment by wojtek-t** [comment]
> Update the release note to describe what people using endpoints or configmaps as their lock resource should do (the specific versions of client-go that support the joint lock functionality).

@liggitt - thanks; updated - PTAL if the current wording looks ok to you

File: staging/src/k8s.io/client-go/tools/leaderelection/resourcelock/interface.go [hunk 1]
```diff
@@ -190,21 +174,15 @@ func New(lockType string, ns string, name string, coreClient corev1.CoreV1Interf
 	}
 	switch lockType {
 	case endpointsResourceLock:
-		return nil, fmt.Errorf("endpoints lock is removed, migrate to %s", EndpointsLeasesResourceLock)
+		return nil, fmt.Errorf("endpoints lock is removed, migrate to %s", endpointsLeasesResourceLock)
 	case configMapsResourceLock:
-		return nil, fmt.Errorf("configmaps lock is removed, migrate to %s", ConfigMapsLeasesResourceLock)
+		return nil, fmt.Errorf("configmaps lock is removed, migrate to %s", configMapsLeasesResourceLock)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/leaderelection/resourcelock/interface.go:179]: this message now points people at something unsupported by the current client-go version... maybe include the recommended client-go version for this in this message (v0.27.x)?

Comment (wojtek-t) [reply staging/src/k8s.io/client-go/tools/leaderelection/resourcelock/interface.go]: done - PTAL

**Pull request merged**